### PR TITLE
feat(terminal): wire Terminal pages to Fly Sprites

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

Wires the existing Terminal page type (full TUI already in place) to Fly Sprites microVMs so commands actually execute instead of returning a hardcoded placeholder.

### What ships

**PR1 — Terminal session core** (pure functions, no IO)
- `deriveTerminalSessionKey` — HMAC-SHA3-256 keyed on `(tenant, drive, page)` + server secret; namespace-versioned to prevent cross-surface collisions
- `planTerminalLifecycle` — pure lifecycle planner (create / resume / deny / teardown); re-authz on every resume so a revoked member can't reconnect to a warm VM
- `acquireTerminalSandbox` — IO orchestrator with injected deps (store, client, clock, secret); no-orphan guarantee on teardown
- 18 unit tests, all pure/injectable

**PR2 — Execute API route**
- `terminal_sessions` table — `pageId` FK (cascade delete), unique `sessionKey`, `userId` FK; migrations 0166 + 0167
- `createDbTerminalSessionStore` — lazy-imported Drizzle adapter (upsert on `sessionKey` conflict)
- `POST /api/pages/[pageId]/terminal/execute` — full pipeline: kill-switch → session-only auth → `canPrincipalEditPage` → body validation → DB lookups → `checkCodeExecutionQuota` → concurrency slot → `acquireTerminalSandbox` → `sprite.runCommand({cmd:'sh', args:['-c', cmd]})` → `truncateToBytes` (64KB cap) → `chargeCodeExecutionBudget` + `auditRequest` + `writeCodeExecutionAudit` (all fire-and-forget) → JSON response; `deny→403`, unexpected errors→500 (catch), slot always released (finally)
- `@fly/sprites` loaded via dynamic `import()` inside the handler (ESM-only package — avoids webpack bundling at build time, same pattern as `sandbox-tools-runtime.ts`)
- 19 route tests covering all paths

**PR3 — UI wiring**
- `TerminalView.handleCommand` → async POST to `/api/pages/[pageId]/terminal/execute`
- `isExecuting` state blocks input while command is in-flight
- Status-code mapping: 401→"Session expired — please reload", 403→"Permission denied", 429→"Quota exceeded", 503→"Code execution disabled", default→"Error {status}: {message}"
- Functional `setSession` to avoid stale history closure

### What this enables
Terminal pages are now real stateful containers: one warm Fly Sprite per page, all editors share the same shell, permissions come from existing RBAC.

### What's deferred
- Streaming output (long-running commands wait synchronously)
- Idle reaper for terminal sessions (TTL-based cleanup is a follow-on)
- GitHub egress / npm access (starts locked-down, widening is a config change)
- Visual "running" indicator beyond blocking input
- Reconnection if a Sprite is deleted server-side between acquire and execute (edge case; platform-level replacement is a follow-on)

## Test plan
- [ ] `bun run typecheck` passes from monorepo root
- [ ] `bun vitest run packages/lib/src/services/sandbox/__tests__/terminal-session-manager.test.ts` — 18/18
- [ ] `bun vitest run apps/web/src/app/api/pages` — 19/19
- [ ] `bun run db:migrate` applies 0166 + 0167 cleanly
- [ ] Create a Terminal page → run `echo hello` → output shows `hello`
- [ ] Run `ls /` → lists root filesystem
- [ ] Session persists across page reload (same Sprite resumes)
- [ ] Viewer cannot execute (403 returned, "Permission denied" shown)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UZJ5pz4gfKLEYvBzTboxqq

### CODE REVIEW

File: apps/web/src/app/api/pages/[pageId]/terminal/execute/route.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  createDbTerminalSessionStore,
+} from '@pagespace/lib/services/sandbox/terminal-session-manager';
+import {
+  checkCodeExecutionQuota,
+  acquireCodeExecutionSlot,
+  releaseCodeExecutionSlot,
+  chargeCodeExecutionBudget,
+} from '@pagespace/lib/services/sandbox/quota';
+import { truncateToBytes } from '@pagespace/lib/services/sandbox/output-limit';
+import { createSpritesSandboxClient } from '@pagespace/lib/services/sandbox/sandbox-client/sprites';
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/pages/[pageId]/terminal/execute/route.ts]: Fixed in commit `36cadedde`. Replaced the static import with a dynamic `import('@pagespace/lib/services/sandbox/sandbox-client/sprites')` inside the handler, exactly following the pattern in `sandbox-tools-runtime.ts`. The module now resolves at call time (after the kill-switch check), so no evaluation happens until code execution is actually enabled and requested.

File: apps/web/src/app/api/pages/[pageId]/terminal/execute/route.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      maxBytes: MAX_OUTPUT_BYTES,
+    });
+
+    const rawOutput = result.stdout + (result.stderr ? '\n' + result.stderr : '');
+    const { text: output } = truncateToBytes({ text: rawOutput, maxBytes: MAX_OUTPUT_BYTES });
+    const durationMs = Date.now() - startMs;
+
+    // Charge the sliding-window budget. Fire-and-forget: a charge failure should
+    // not cause the command output to be discarded (the command already ran).
+    chargeCodeExecutionBudget({ userId, driveId, tenantId }).catch(() => {});
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/pages/[pageId]/terminal/execute/route.ts:160]: The fire-and-forget charge is deliberate — a charge failure must not discard output the command already produced (the same reasoning applies here as to other fire-and-forget side effects in this route). The advisory `checkCodeExecutionQuota` preflight gates each request; the charge records usage in the sliding window so *subsequent* requests enforce the budget. This matches the intentional design of the quota system and is consistent with how `sandbox-tools-runtime.ts` wires the charge for the agent code-execution tools. Quota window over-runs (concurrent requests racing the preflight) are bounded by both the per-user concurrency slot limit (step 7) and the advisory check — the exposure window is small and accepted.

File: packages/lib/src/services/sandbox/terminal-session-manager.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        }
+        await safeTouch(deps.store, key, deps.now());
+        return { ok: true, sandboxId: handle.sandboxId, resumed: true };
+      }
+
+      case 'teardown': {
+        const stopped = await safeStop(deps.client, plan.sandboxId);
+        if (!stopped) {
+          // VM may still be running — keep the link so the idle reaper can reclaim it.
+          return { ok: false, reason: 'error' };
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/services/sandbox/terminal-session-manager.ts:220]: Valid edge case — a platform-deleted Sprite that passes the idle-timeout check will fail on `safeStop` and loop returning 500. Deferred because the PR description explicitly calls out 'Idle reaper for terminal sessions (TTL-based cleanup is a follow-on)'. The teardown path is unreachable until an idle-timeout value is wired in (`idleTimeoutMs` defaults to undefined, so `planTerminalLifecycle` always returns `create` or `resume`, never `teardown`). The fix (delete the stale DB row on a 404 stop) will land with the idle-reaper follow-up where the teardown path becomes active.
